src/asl.ts

```diff
@@ -1955,19 +1953,45 @@ export class ASL {
     }
 
     if (isTemplateExpr(expr)) {
-      return this.evalContext(expr, ({ evalExprToJsonPathOrLiteral }) => {
+      return this.evalContext(expr, ({ evalExpr, addState }) => {
         const elementOutputs = [
           {
             value: expr.head.text,
             containsJsonPath: false,
           },
-          ...expr.spans.flatMap((span) => [
-            evalExprToJsonPathOrLiteral(span.expr),
-            {
-              value: span.literal.text,
-              containsJsonPath: false,
-            },
-          ]),
+          ...expr.spans.flatMap((span) => {
+            const out = evalExpr(span.expr);
+
+            /**
+             * It is possible that expressions in the template can update variables used in the template.
+             * For JsonPath outputs assign the output of any json path to a new heap variable
+             * instead of whatever json path they return. Do the same for conditions, they need have a boolean assigned
+             * to a heap variable anyways. We can handle jsonPath and conditions in the same way.
+             *
+             *
+             * ```ts
+             * let x = "1";
+             * ${x} ${(x = "2")}` // should output `1 2`
+             * ```
+             *
+             * If the heap assignment isn't necessary, we will optimize out the extra assignment later.
```

Comment:
Does this "later" code exist already? Can you link to where this will be done?

Reply:
it does not exist yet, wanted to get everything lined up how I wanted it and then write the optimization. Doing to re-write to execute in order now and will write the optimization after.

Reply:
https://github.com/functionless/functionless/issues/415

Reply:
Hmm, can you explain how inefficient the generated ASL will be if we merge this?

Reply:
> Hmm, can you explain how inefficient the generated ASL will be if we merge this?

It will work correctly for all cases. 

Each dynamic assignment in an object, array, or template literal will create an extra state and variable.

(This pr is still draft so I can take a crack at the optimization).

Reply:
How may states doe this require?

```ts
{
  constant: "value",
  dynamic: x
}
```